GPII-1897: Isolate buildup and teardown to allow for persistence between runs.

the-t-in-rtf opened the pull request:
See [GPII-1897](https://issues.gpii.net/browse/GPII-1897) for details.

the-t-in-rtf committed:
NOJIRA:  Updated linting infrastructure.

the-t-in-rtf committed:
NOJIRA:  Updated gpii-express version.

the-t-in-rtf committed:
GPII-1897: Break apart existing express pouchdb grades and isolate buildup and teardown cycle.

the-t-in-rtf committed:
GPII-1897:  Updated documentation to reflect recent work.

the-t-in-rtf committed:
GPII-1897:  Updated JSDocs.

the-t-in-rtf commented:
Even with my best efforts, I cannot seem to get the in-memory `memdown`-backed grade to consistently clear itself out.  I get constant errors like:

```
{ [conflict: Document update conflict]
  status: 409,
  name: 'conflict',
  message: 'Document update conflict',
  error: true }
```

I only get the errors when loading test data, likely these are records with hard-coded `_id` values, which cannot be overridden (records without would silently result in duplicates).

We should look at this as part of this PR.  It may be that we can mitigate this by removing in-memory storage as an option for express-pouchdb, and only allowing in-memory databases on the browser side.  Now that the filesystem-backed grade has been better fleshed out, I'm happy only using that in my own work.

the-t-in-rtf committed:
GPII-1897:  Continued hardening for both "persist" and "reload" test cases.

the-t-in-rtf committed:
GPII-1896: Documentation, initial implementation, and tests for cross-platform PouchDB wrapper grade.

the-t-in-rtf committed:
GPII-1896: Added additional tests for PouchDB component.

the-t-in-rtf committed:
GPII-1896: Removed `inMemory` grades.

the-t-in-rtf committed:
GPII-1897: Added node-specific loader for gpii.pouch (available in gpii.pouch.node).

the-t-in-rtf committed:
GPII-1896:  Added Testem configuration for browser tests.

the-t-in-rtf committed:
GPII-1897: Integrated new `gpii.pouch` component into `gpii.pouch.express` grade.

the-t-in-rtf commented:
The new `gpii.pouch` grade is now integrated into `gpii.pouch.express`.  This should be ready for further review, and for Cindy and Simon to examine.

the-t-in-rtf committed:
NOJIRA:  Minor formatting changes in docs.

the-t-in-rtf committed:
GPII-1897:  Promoted "harness" grades to non-test namespace per feedback from the team.

the-t-in-rtf committed:
GPII-1988: Documented QI config files with added comments.

the-t-in-rtf committed:
NOJIRA:  Minor update of dependencies for latest PR, based on output of `npm outdated`.

amb26 commented on src/js/pouch-express.js:
```diff
@@ -0,0 +1,287 @@
+        dbComponentOptions.dbPaths = expandedDef.data;
+    }
+
+    var dbComponent = fluid.component(dbComponentOptions);
```
This must never be done - the fact that it still works is just the blindest accident and will be cleaned up in the next framework update. Always use initDependent (if you must have a programmatic API) or preferably the Nexus methods such as fluid.construct

the-t-in-rtf replied:
The programmatic API was used to support the previous `databases` option rather than forcing everyone to define subcomponents for each database.  I will try the Nexus methods here and we can review later.

the-t-in-rtf replied:
Done, I migrated to using `fluid.construct`.

the-t-in-rtf replied:
Looking at this in light of other comments, `dbOptions` can easily become the basis of the options set for the "array of dynamic components" we've been discussing.

the-t-in-rtf committed:
GPII-1897:  Migrated to using `fluid.construct` to construct databases per PR feedback.

the-t-in-rtf commented:
This needs to be merged after [gpii-express](https://github.com/GPII/gpii-express/pull/25), so that we can move to a published version.

amb26 commented on docs/harness.md:
```diff
@@ -0,0 +1,26 @@
+
+| Option               | Type       | Description |
+| -------------------- | ---------- | ----------- |
+| `baseDir` (required) | `{Number}` | A full or package-relative path to the base directory in which all database content, configuration files, and logs will be stored. |
```
Type of {Number} doesn't seem right here

the-t-in-rtf replied:
Fixed.

amb26 commented on docs/pouch-express.md:
```diff
@@ -0,0 +1,129 @@
+## Component Invokers
+
+### `{that}.cleanup()`
+* Returns: A `Promise` that will be resolved once cleanup is complete (the `onCleanupComplete` event should also be fired.
```
Missed parenthesis

the-t-in-rtf replied:
Fixed.

amb26 commented on docs/pouch-express.md:
```diff
@@ -0,0 +1,129 @@
+overriden with another invoker.
+
+### `{that}.middleware(request, response, next)`
+* `request`: The [request object](http://expressjs.com/en/api.html#req) provided by Express, which wraps node's [`http.incomingMessage`](https://nodejs.org/api/http.html#http_class_http_incomingmessage).
```
I've commented before on this copypasta language from gpii-express - this isn't an example of "wrapping", it simply is an instance of http.incomingMessage

the-t-in-rtf replied:
Sorry, didn't catch the wording here in the last revamp. It's fixed now.

amb26 commented on docs/pouch-express.md:
```diff
@@ -0,0 +1,129 @@
+## Component Invokers
+
+### `{that}.cleanup()`
+* Returns: A `Promise` that will be resolved once cleanup is complete (the `onCleanupComplete` event should also be fired.
```
Paren - also "will" inspires more confidence than "should"

the-t-in-rtf replied:
Agreed, I have purged all "shoulds" throughout.

the-t-in-rtf replied:
Also, fixed the paren.

amb26 commented on docs/tests.md:
```diff
@@ -62,15 +62,15 @@ fluid.defaults("my.tests.environment", {
-equivalent to CouchDB.  Although you will generally use this in combination with the test enviroment above, it can
-also be used on its own. As an example, the file `tests/js/launch-test-harness.js` included with this package can be
-used to launch a standalone test instance of this package for manual QA.
+An extension of `gpii.tests.pouch.environment`, which allows for cases in which data should persist between runs.
```
This should read gpii.test.pouch.environment

the-t-in-rtf replied:
I had to create one at least to test persistence in this package.  I can see moving it into those tests and out of the documentation, if that seems reasonable.

the-t-in-rtf replied:
I collapsed this into the environment in the persistence tests.  It is no longer exposed via loadTestingSupport (or documented).

amb26 commented on docs/tests.md:
```diff
@@ -62,15 +62,15 @@ fluid.defaults("my.tests.environment", {
-equivalent to CouchDB.  Although you will generally use this in combination with the test enviroment above, it can
-also be used on its own. As an example, the file `tests/js/launch-test-harness.js` included with this package can be
-used to launch a standalone test instance of this package for manual QA.
+An extension of `gpii.tests.pouch.environment`, which allows for cases in which data should persist between runs.
```
Seems awkward as a capability - there shouldn't be a use case for test cases which persist data between runs. We _do_ have a requirement for a pouch configuration that persists data, but I don't see that this needs to be packaged as a testEnvironment?

amb26 commented on package.json:
```diff
@@ -5,21 +5,26 @@
     "infusion": "2.0.0-dev.20160519T222603Z.754d2c6",
-    "istanbul": "0.4.4",
+    "istanbul": "0.4.5",
+    "lie": "3.1.0",
```
I don't see that this dependency is used

the-t-in-rtf replied:
It's used when `npm test` is run, see:

https://github.com/the-t-in-rtf/gpii-pouchdb/blob/GPII-1897/package.json#L10

the-t-in-rtf committed:
NOJIRA:  Updated QI config to use "lts" version of node.

amb26 commented on src/js/pouch-express.js:
```diff
@@ -0,0 +1,288 @@
+
+/*
+
+    TODO:  Work through the express-pouchdb setup and code with Antranig, as we see many warnings like:
```
I imagine this is a bug in express-pouch, given the insufferable bureaucracy of bluebird-promise. It seems that one way we could trace it down is to patch inside express-pouch to include var Promise = require("./js/debug/bluebird"); as per https://github.com/petkaantonov/bluebird/issues/1205 and see where it is.

amb26 commented on src/js/pouch-express.js:
```diff
@@ -0,0 +1,288 @@
+    tmpDir:  os.tmpdir(),
+    baseDir: "@expand:path.resolve({that}.options.tmpDir, {that}.id)",
+    expressPouchConfigFilename: "config.json",
+    expressPouchConfigPath:     "@expand:path.resolve({that}.options.baseDir, {that}.options.expressPouchConfigFilename)",
```
Slightly puzzled and a bit terrified that this works, mostly likely due to some corruption of the global object in which Infusion is mounted.

the-t-in-rtf replied:
Not sure why this would be any more terrifying than calling a jqUnit function from a declarative block.  If it's a risk, I suppose I could make a small function, but it seems a waste.

amb26 replied:
It's more terrifying because there is a good reason for jqUnit to be in Infusion's global namespace since we explicitly load it there. "path" on the other hand could only be there because it just so happens that part of Infusion's boot loader code happens to need it

the-t-in-rtf replied:
Thanks for explaining.  To confirm, the preferred approach is to ensure that the component  actually requires path directly?

the-t-in-rtf replied:
Anyway, there is now a named function used in the expander, so this has been addressed.

the-t-in-rtf replied:
I was mistaken and it's still needed.  I put the require back in for now.

amb26 commented on tests/js/pouchdb-component-tests/pouchdb-component-common-tests.js:
```diff
@@ -0,0 +1,354 @@
+};
+
+// A simple function to test the support for "reduce" functions using the query method.
+gpii.tests.pouchdb.component.common.reduce = function (keys, values, rereduce) {
```
Could you explain a little more clearly what it does, and what its arguments are?

the-t-in-rtf replied:
I added links to the relevant CouchDB documentation and briefly explained the output of both the map and reduce functions.

amb26 commented on src/test/caseHolder.js:
```diff
@@ -0,0 +1,37 @@
+
+fluid.registerNamespace("gpii.test.pouch.caseHolder");
+
+gpii.test.pouch.caseHolder.standardSequenceEnd = [
```
Give this a more descriptive name rather than merely "standard"

the-t-in-rtf replied:
Done, I added a courtesy comment as well to explain what was being done.

amb26 commented on src/js/pouchdb-node.js:
```diff
@@ -0,0 +1,169 @@
+
+    var expandedPaths = gpii.pouch.node.expandPath(dbPath);
+    fluid.each(expandedPaths, function (dbPath) {
+        if (dbPath) {
```
Why might we have a dbPath that is empty, in a condition which is not an error?

the-t-in-rtf replied:
There might simply be no data to load, as is the case with the "noData" example in the tests, but I believe those are screened out in the code that calls loadDataFromPath.  Will confirm.

the-t-in-rtf replied:
I moved the check outwards and added a comment to clarify.

amb26 commented on src/js/pouchdb-node.js:
```diff
@@ -0,0 +1,169 @@
+ *
+ */
+gpii.pouch.node.expandPath = function (dbPath) {
+    if (Array.isArray(dbPath)) {
```
This body (or even function) can just be replaced by fluid.makeArray

the-t-in-rtf replied:
Done and all references updated.

amb26 commented on src/js/pouchdb-node.js:
```diff
@@ -0,0 +1,169 @@
+ * @param that - The component itself.
+ *
+ */
+gpii.pouch.node.initDir = function (that) {
```
This function can be replaced by a call to https://github.com/substack/node-mkdirp

the-t-in-rtf replied:
We should discuss further, as this is deeply entangled with the "do I already have data" check below.

amb26 commented on src/js/pouchdb-node.js:
```diff
@@ -0,0 +1,169 @@
+
+    if (fs.existsSync(that.options.baseDir)) {
+        if (fs.existsSync(fullPath)) {
+            that.hasExistingDataDir = true;
```
Is this flag used or documented?

the-t-in-rtf replied:
Yes, this is the means by which I guess at whether there is already data, to satisfy one of Cindy's requirements.  It is used in "loadDataIfNeeded":

https://github.com/the-t-in-rtf/gpii-pouchdb/blob/GPII-1897/src/js/pouchdb-node.js#L88

We should review the PouchDB and/or leveldown docs collectively, there must be a clearer and cleaner way to do this.

amb26 commented on src/js/pouchdb-common.js:
```diff
@@ -0,0 +1,145 @@
+        skip_setup: true
+    },
+    events: {
+        onAllDocsComplete: null,
```
I think there's a reasonable argument for replacing all of these with a single event e.g. onPouchOperationComplete with a suitably wrapped argument payload

the-t-in-rtf replied:
Please explain further, it seems much safer to me to have separate events.  Having a single event seems like it would require people to write listeners that include code to filter by sub-event.

amb26 replied:
I'd written a response that read like this: 

In fact this only gives an impression of safety. Replacing a promise-based API one for one with an event-based one creates an unavoidable risk of "cross-contamination" between requests. Having more event types reduces the contamination but does not eliminate it. So we may as well be honest that these events are only intended for the lightweight use of the IoC testing framework. I've written up https://issues.fluidproject.org/browse/FLUID-5949 to describe these kinds of risks further, and propose an improvement in the testing framework that will remove the encouragement to create these kinds of event-based wrappers.

I thought further, and realised that I don't fully believe in this advice. There _do_ seem to be some legitimate use cases for allowing the user to intercept these kinds of events by type, for example, to interpose special behaviour or processing for example, on "onCompactComplete" etc.

However, you will see my point about the risks of contamination - there's no way that the consumer of the onGetComplete event can know _which_ get is complete. There is nothing which associates an particular invocation of {pouch}.get with a particular firing of {pouch}.events.onGetComplete - in test sequences we tend to fudge by assuming that the _next_ invocation of onGetComplete has to be the matching one, but it's clear that in real use this is faulty.

So what I propose instead is this - that you eliminate all of these invokers and all of these events, and move instead to the "transforming promise sequence" pattern that we see in DataSource. "get" is now signalled by that.get.fire, and the "implementation" of the invoker just becomes another listener to this event. It's then possible i) for users to interpose both before and after the actual dispatch to Pouch, and ii) the ground is opened for a future generic "request id" system that dumps an extra value into the "options" structure of fluid.promise.sequence iii) we no longer have the annoying pleonasm of a whole raft of invokers and events with corresponding but not completely matching names

In practice we plan to abolish invokers completely, and we may as well start experimenting with this technique in earnest and develop infrastructure to support it. I will also try to take care of https://issues.fluidproject.org/browse/FLUID-5949 as soon as possible

the-t-in-rtf replied:
Thanks for the explanation.  This seems a bit large to include in this particular PR.  I propose creating a ticket for the work, to be discussed in our regular meeting and begun once we have even an early mechanism to support promises in Fluid IoC tests.

the-t-in-rtf replied:
See [GPII-1995](https://issues.gpii.net/browse/GPII-1995).

amb26 commented on src/js/pouch-express.js:
```diff
@@ -0,0 +1,288 @@
+        // We cannot simply call db.destroy() on all databases, because express-pouchdb creates a few databases on its own
+        // and gives us no way to clean those up.
+        //
+        // The surest approach is to completely remove the underlying directory.  However, express-pouchdb will have one
```
Seems to me that there's little chance this dodge will work on Windows. Perhaps we're reaching the midpoint of an argument that it's time to throw away express-pouchdb?

the-t-in-rtf replied:
We would need something very much like express-pouch even for basic dataSource tests.  Creating something ourselves that covered basic tests is possible, but a fair amount of effort.  

It's also the only way we have of testing search integration with CouchDB.  Creating something that would allow us to test our search integration would be a monumental task.

We should discuss other options in the next meeting.  I have considered instrumenting the PouchDB builder so that it has an awareness of what databases exist, regardless of who created them.  That would give us the handles we need to clean up properly.

We could also create our own instrumented adapter based on leveldown, and pass it in as a dbOption.  (We do this with memdown in the cleanup code).  This approach might also give us enough information about their adapter API to create a "noop" adapter that we can use instead of memdown during the cleanup phase.

the-t-in-rtf replied:
This is what I mean: https://pouchdb.com/adapters.html

We would use the default disk-based adapter as our starting point.

amb26 commented on src/js/pouch-express.js:
```diff
@@ -0,0 +1,288 @@
+    //   2. A way to clean up each database later on.  We do this by retaining a list of database instances.
+    //   3. A way to know when all of them have been destroyed.  We do this with another `sequence` in our cleanup invoker (see below).
+    //
+    // I can think of ways to accomplish #2 with dynamic components, but not 1 and 3.
```
The use of dynamicComponents would be more obviously declarative. This can be done via the "array" style of dynamicComponents sourced from that.options.databases. http://docs.fluidproject.org/infusion/development/SubcomponentDeclaration.html#dynamic-subcomponents-with-a-source-array You can then pick them up later using fluid.queryIoCSelector http://docs.fluidproject.org/infusion/development/IoCAPI.html#fluid-queryiocselector-root-selector-flat- and start the asynchronous operations after onCreate etc and

the-t-in-rtf replied:
Just to confirm, I can reference material within a `{source}` in the usual manner?  For example:

```
fluid.defaults("examples.colorizer", {
    gradeNames: ["fluid.component"],
    values: [{ color: "red", transparency: 50}, {color: "blue", transparency: 25 }],
    dynamicComponents: {
        dynamic: {
            sources: "{that}.options.values",
            type: "fluid.component",
            options: {
                color: "{source}.color",
                transparency: "{source}.transparency"
            }
        }
    }
});
```

Please confirm.

the-t-in-rtf replied:
OK, I have thought through most of this, I have a clear sense of what options to use and how to sequence the destroy promises and then fire an event.  How do I ensure that gpii.pouch.express fires an event when all dynamic components are finished being created?  That's the only piece I'm missing as far as I know.

amb26 replied:
i) Yes, you can reference material in {source} - https://github.com/fluid-project/infusion/blob/master/tests/framework-tests/core/js/FluidIoCTests.js#L3048
ii) I suggest you just move this to onCreate - since the current framework discourages the use of events before component construction is complete

the-t-in-rtf replied:
To confirm, are you saying that onCreate for the parent component will only be fired once the dynamic components have been created?  If so, I have what I need to make this change.

amb26 replied:
Yes, that's correct

the-t-in-rtf replied:
I am looking into this further.  It's not whether they've been created that I care about, it's whether their data has been loaded.  How can I keep track of that?

amb26 replied:
Add a listener during onCreate that queues some promise member of the components into a sequence, and then fires an event

the-t-in-rtf committed:
GPII-1897: Removed extraneous function in favor of fluid.makeArray.

the-t-in-rtf committed:
GPII-1897:  Disabled Safari in the default testem runs, as it causes the tests to hang on completion.

the-t-in-rtf committed:
GPII-1897:  Fixed various typos and weak language in the documentation.

the-t-in-rtf committed:
GPII-1897:  Moved the "persistent test harness" out of the documentation and publicly exposed fixtures and into the persistence tests.

the-t-in-rtf committed:
GPII-1897:  Improved documentation for test map/reduce functions.

the-t-in-rtf committed:
GPII-1897:  Renamed standard end sequence steps and added a comment to explain their purpose.

amb26 commented on src/js/pouchdb-node.js:
```diff
@@ -0,0 +1,145 @@
+    // This instance may not actually have any data, but it should still incidate that it has finished (not) loading data.
+    if (dbPaths !== undefined) {
+        fluid.each(fluid.makeArray(dbPaths), function (singleDbPath) {
+            var data = fluid.require(fluid.module.resolvePath(singleDbPath));
```
The original fluid.require should have been sufficient here

the-t-in-rtf replied:
That crept in when I removed the previous path resolution function.  I have removed the unneeded call to `fluid.module.resolvePath`.

the-t-in-rtf committed:
GPII-1897:  Removed unneeded call to `fluid.module.resolvePath` within call to `fluid.require`.

the-t-in-rtf committed:
GPII-1897:  Standardized documentation for `request`, `response`, `next`, etc. arguments with links to the relevant `gpii-express` documentation.

cindyli commented on tests/js/pouchdb-component-tests/pouchdb-component-common-tests.js:
```diff
@@ -0,0 +1,356 @@
+                name: "Test the `PUT` method (with an ID)...",
+                sequence: [
+                    {
+                        func: "{testEnvironment}.pouchDb.post",
```
This test is to test the `PUT` method but still calling `post()` API. `POST` has already been tested above. a mistake from a copy-paste? :)

the-t-in-rtf replied:
This is fixed in a later commit.

the-t-in-rtf replied:
Hi, Cindy.  Can you somehow indicate that the change has been made?  I don't see an option to do anything about the request from this end.

cindyli replied:
There's an "approve changes" link at my end. Resolved.

the-t-in-rtf committed:
GPII-1897: Refactored to eliminate 409 errors caused by PouchDb's destroy function.

the-t-in-rtf committed:
GPII-1897:  Fixed PUT test to actually use the PUT method.

the-t-in-rtf commented:
In working to simplify the express pouchdb cleanup cycle, I traced most of the previous 409 errors to the PouchDB destroy method.  I wrote a safe alternative based on previous approaches (bulk get, flag all records as deleted, bulk update, compact).  This gets rid of all but the 409 errors in the “persistence” tests (which Antranig and I can review further later).

As discussed, I also compared “modes” that enable/disable various CouchDB work-alike functions.  The default is “full”, which we use at the moment.  “minimum” works with our tests, but is not appreciably faster, and does not avoid the 409 errors we saw.

With the “custom” mode, you can cherry-pick particular features you need and disable the rest.  I disabled replication, authentication, and nearly every other feature that we don’t use.  Even with that, we still saw 409 errors and the speed was actually a bit slower, but still within a few hundredths of a second of 3 seconds per run.

The good news is that in doing this research I figured out the right options to disable the “changes” API, which was responsible for leaked listeners.

I had discussed writing a custom adapter, but I realized I could do everything I needed if I simply added the three databases that express-pouchdb creates to our databases option.

the-t-in-rtf committed:
GPII-1897:  Reverted previous removal of `path` include.

cindyli commented on src/js/pouch-express.js:
```diff
@@ -0,0 +1,269 @@
+ * @returns {Promise} - A promise that will be resolved when cleanup is complete.
+ *
+ */
+gpii.pouch.express.cleanup = function (that) {
```
Do you plan to continue to return a promise for this API function? It's now not returning anything.

the-t-in-rtf replied:
I can add that back, no problem.

cindyli commented:
@the-t-in-rtf, I've tried the refactored cleanUp() API. Seems it no longer removes the physical database on the hard disk. This leaves lots of strayed db files on the system temp directory after running node db tests.

The good news is 409 error is gone when running tests. I believe it's because each test sequence is using its own individual physical db files.

In the case that the same physical file location is used for the end point of /reset-pouch, two issues:
1. cleanUp() no longer returns a promise (have commented earlier). It would be great if returning a promise could be maintained. It's pretty handy for further operations.
2. After the cleanUp(), the restart of the pouchHarness doesn't reload the data. Viewing the database via Futon shows that old records were marked as "_deleted: true" while new records don't exist.

Note that the above test with /reset-pouch was using the existing approach of restarting pouchHarness. I will hold off on refactoring /reset-pouch for now (thinking of replacing the restart of pouchHarness with a call to initDbs()) until I hear back from you or cleanUp() is more stable.

Thanks.

cindyli commented:
Noticed there are also physical db files left behind with the old cleanUp() after running tests. So, now I'm not sure if it's the cleanUp() that didn't remove those files. It could be my scripts that didn't call cleanUp() when needed. I will have a closer look tomorrow.

cindyli commented:
@the-t-in-rtf, confirmed that the new implementation of cleanUp() no longer removes physical db files from the hard disk with the good news of 409 error being gone.

I did have some db tests not having cleanUp() being called in between test sequences. That's why there were always some physical db files left behind when running all-tests. I will fix that in my branch.

the-t-in-rtf commented:
Hi, @cindyli.  The new approach should leave temporary content in place, but was meant to result only in empty, compacted database files, which should be acceptably small. Apparently compaction wasn't enough to actually remove the data, I will have to test a bit further before our next meeting.

the-t-in-rtf committed:
GPII-1897:  Made individual `gpii.pouch.node` instances responsible for their own cleanup, including removing FS content.  Added code to optionally remove FS content.

the-t-in-rtf committed:
GPII-1897:  Updated "is there data?" check to use db.info().

cindyli commented on src/js/pouchdb-node.js:
```diff
@@ -0,0 +1,194 @@
+var path   = require("path");
+
+var rimraf = require("rimraf");
+var mkdirp = require("mkdirp");
```
Please include the dependency on `mkdirp` into package.json.

the-t-in-rtf replied:
I have it, just testing some other updates before committing.  Sorry.

cindyli commented on src/js/pouchdb-node.js:
```diff
@@ -94,10 +98,53 @@ gpii.pouch.node.loadDataIfNeeded = function (that) {
+                }
+                else {
+                    if (that.options.removeDirOnCleanup) {
+                        rimraf(that.options.dbOptions.prefix, function (rmDirError) {
```
This recursively removes everything in the directory of `that.options.dbOptions.prefix`, which in this case is everything in the system temp directory, instead of only the database specific directory.

cindyli replied:
This is an issue when using `gpii.pouch.harness.persistent` where `that.options.dbOptions.prefix` equals to the `baseDir` specified in that grade instead of the individual database directory.

the-t-in-rtf replied:
so it should remove prefix + dbName?

cindyli replied:
Yes.

the-t-in-rtf replied:
I've just fixed this and will commit shortly.

cindyli commented on src/js/pouchdb-node.js:
```diff
@@ -0,0 +1,194 @@
+    },
+    events: {
+        onDataLoaded:null,
+        onDirExists: null,
```
This event is not used anywhere.

the-t-in-rtf replied:
Thanks, Cindy.  This is fixed now.

cindyli commented:
To summarize the result of trying out the latest changes with cleanup() and data reload:

- When using cleanup() alone at running tests, cleanup() works great with removing physical individual db directory. SUPER! Thanks. (BTW, Please keep in mind to return a promise for this API. Thanks.)

- In the circumstance of cleanup() + data reload by re-instantiating pouchHarness, which is used by the endpoint /reset-pouch to reset pouchDB to its initial data set. 2 issues:
1. The entire OS temp directory is cleaned up, meaning everything in that directory was removed, not only the specific pouchDB directory;
2. Futon shows the data has been reset to the initial data set, however, the database seems pointing to a replication of the old database because old data changes are still retrievable. Restarting the entire GPII node instance will refresh the backend to point to the reset database. Of course, restarting GPII is not desired as that's what the endpoint of /reset-pouch is created for.

the-t-in-rtf commented:
Just to summarize, the recent concerns around both the filesystem cleanup and the initialization checks have been addressed.  There is still some work to be done regarding our use of promises and the creation of dynamic components.

the-t-in-rtf committed:
GPII-1897:  Better targeted cleanup routine to focus on individual database directories.  Also removed unneeded event.

the-t-in-rtf committed:
GPII-1897:  Updated all dependencies to eliminate security vulnerabilities detected by `snyk`.

the-t-in-rtf commented:
I believe I've addressed all of the concerns @cindyli raised except for the Futon issues.  I suspect we'll need a more general cleaner to go with the new loader discussed in [GPII-1987](https://issues.gpii.net/browse/GPII-1987).  The cleaner can use [the purge endpoint](http://docs.couchdb.org/en/2.0.0/api/database/misc.html#db-purge) to ensure the deleted data is well and truly gone.

the-t-in-rtf committed:
GPII-1897:  Returned promise from cleanup function.

cindyli commented on docs/pouchdb.md:
```diff
@@ -0,0 +1,199 @@
+| `onBulkDocsComplete`         | Fired when a call to `{that}.bulkDocs` (see below) is successfully completed. |
+| `onBulkGetComplete`          | Fired when a call to `{that}.bulkGet` (see below) is successfully completed. |
+| `onCompactComplete`          | Fired when a call to `{that}.compact` (see below) is successfully completed. |
+| `onDestroyPouchComplete`     | Fired when a call to `{that}.destroyPouch` (see below) is successfully completed. |
```
Remove this event as it's no longer available with the removal of `destroyPouch()` from `pouchdb-common.js`.

the-t-in-rtf replied:
Thanks, Cindy.  Fixed!

the-t-in-rtf committed:
GPII-1897:  Removed docs reference to outdated "destroy" invoker and event.

the-t-in-rtf committed:
GPII-2107: Added a well-tested mechanism to safely recreate express-pouchdb without restarting.

the-t-in-rtf commented:
Once Cindy confirms the fix for [GPII-2107](https://issues.gpii.net/browse/GPII-2107), we should agree on what (if anything) needs to be done before this can be merged.  AFAIK the open topics are:

1.  Refactoring to use promises instead of events.
2. Using the "array" mechanism to create the database components.

If possible, I would rather create a new ticket for each, as I am unlikely to have time to work on them immediately, and they are not currently blocking anyone's work.

the-t-in-rtf committed:
GPII-1897:  Collapsed "harness" cleanup into gpii.pouchdb.express.

the-t-in-rtf committed:
GPII-1897:  Added step to recreate the initial data after a successful cleanup.

the-t-in-rtf committed:
GPII-1897:  Added db initialization to cleanup, so that `onCleanupComplete` means that everything is completely reset.

the-t-in-rtf committed:
GPII-1897:  Added view cleanup step after data load.

the-t-in-rtf committed:
GPII-1897: Added code to manually clean up view directories.

the-t-in-rtf committed:
GPII-1897:  Added view requests to persistence tests.

the-t-in-rtf committed:
FLUID-5963: Updated to standard ESLint configuration (and updated dependencies).  Fixed lint errors.

the-t-in-rtf committed:
GPII-1897:  Revert to previous “destroy” approach to cleaning data.

the-t-in-rtf committed:
NOJIRA:  Clean up reports before each npm test run.

the-t-in-rtf commented:
@cindyli, I have completed the refactor we discussed in our last meeting.  We now use the previous "destroy" method to cleanup database instances.  Please confirm if this works for you.

the-t-in-rtf committed:
GPII-1897:  Update the docs to reintroduce the previous “destroyPouch” mechanisms.

the-t-in-rtf committed:
GPII-1897: Updated the docs to remove references to the previous cleanup mechanism and clarify the cleanup behavior in pouch-express.

amb26 commented:
Note that the strategy for sequencing test runs in package.json fails on Windows:     
    "test": "npm run test:browser; npm run test:node",

the-t-in-rtf committed:
GPII-1897: Sidegraded to patched version of express-pouchdb to isolate error logging.

the-t-in-rtf committed:
GPII-1897:  Updated “test” script definition to work on Windows.

the-t-in-rtf commented:
Thanks, @amb26.  Based on the common support of &&, I foolishly assumed support for using semicolons across platforms.  I have updated the definition to use &&, which works on Windows.

the-t-in-rtf committed:
GPII-1897:  Updated to use released version of `gpii-express`.

the-t-in-rtf committed:
GPII-2161: Committed `yarn.lock`.

the-t-in-rtf commented:
To summarize, after much effort the project is in a usable state at the moment.  As mentioned [above](https://github.com/GPII/gpii-pouchdb/pull/12#issuecomment-256929510), I have created issues for the remaining work, which now includes tracking down the underlying source of the 409 errors:

1. [GPII-2162 (document conflicts)](https://issues.gpii.net/browse/GPII-2162)
2. [GPII-1995 (migrating away from event wrappers)](https://issues.gpii.net/browse/GPII-1995)
3. [GPII-2163 (using dynamic components instead of the current `databases` option in gpii.pouch.express)](https://issues.gpii.net/browse/GPII-2163)

@cindyli and @amb26, I hope that a) I have summarized the work remaining correctly and b) we can agree to merge what we have.

amb26 commented on src/js/pouch-express.js:
```diff
@@ -0,0 +1,295 @@
+
+var PouchDB        = require("pouchdb");
+
+var path           = require("path"); // eslint-disable-line no-unused-vars
```
With respect to the discussion below about inclusion of "path", this require is ineffective. All it does is load path into *your* global object, not into Infusion's.
The safe way of ensuring that "path" is available for global resolution is to write
fluid.require("path", require, "path");
http://docs.fluidproject.org/infusion/development/NodeAPI.html#fluid-require-modulename-foreignrequire-namespace-

the-t-in-rtf replied:
Thanks, and fixed.  Given that I had to tiptoe around the ESLint "unused variable" warning, it should be easy enough to remember when this is relevant in the future.

the-t-in-rtf committed:
GPII-2161: Committed `yarn.lock`.

the-t-in-rtf committed:
GPII-1897:  Updated `path` require to use `fluid.require`.
